src/analyses/variable-sensitivity/value_set_abstract_object.cpp

```diff
@@ -130,29 +202,56 @@ abstract_object_pointert value_set_abstract_objectt::expression_transform(
   const abstract_environmentt &environment,
   const namespacet &ns) const
 {
-  std::size_t num_operands = expr.operands().size();
-  PRECONDITION(operands.size() == num_operands);
+  PRECONDITION(operands.size() == expr.operands().size());
 
-  std::vector<abstract_object_sett> collective_operands;
-  collective_operands.reserve(num_operands);
-  for(const auto &op : operands)
-  {
-    auto vsab = std::dynamic_pointer_cast<const value_set_abstract_objectt>(
-      maybe_unwrap_context(op));
-    INVARIANT(vsab, "should be a value set abstract object");
-    collective_operands.push_back(vsab->get_values());
-  }
+  auto collective_operands = unwrap_operands(operands);
+
+  if(expr.id() == ID_if)
+    return evaluate_conditional(
+      expr.type(), collective_operands, environment, ns);
 
   abstract_object_sett resulting_objects;
+
+  auto dispatcher = *values.begin();
   for_each_comb(
     collective_operands,
-    [&resulting_objects, this, &expr, &environment, &ns](
+    [&resulting_objects, &dispatcher, &expr, &environment, &ns](
```

Comment:
Given that it will either be unary or binary, this feels like over-kill.  Is there any reason not to replace this with:

```
if (expr.children() == 1) {
  for (...) { recursively eval }
} else if (expr.children() == 2) {
  for (...) {
    for (...) {
      recursively eval;
    }
  }
```

Reply:
There can be an arbitrary number of operands, yes, and I've added tests explicitly to exercise for cases where there are more than 2.